### PR TITLE
Document Rust's stance on `/proc/self/mem`

### PR DESCRIPTION
**Pull request by sunfishcode**
Add documentation to `std::os::unix::io` describing Rust's stance on
`/proc/self/mem`, treating it as an external entity which is outside
the scope of Rust's safety guarantees.

### CODE REVIEW

File: library/std/src/os/unix/io/mod.rs [hunk 1]
```diff
@@ -44,6 +44,26 @@
...
+//! read from or written to, reads and writes the process's memory. These reads
+//! and writes happen outside the control of the Rust compiler, so they do not
+//! uphold Rust's memory safety guarantees.
+//!
+//! Does this mean that all APIs that might allow `/proc/self/mem` to be opened
+//! and read from or written to must be `unsafe`? No. Rust's safety guarantees
+//! only cover what the program itself can do, and not what entities outside
+//! the program can do to it. `/proc/self/mem` is considered to be such an
+//! external entity, along with debugggers and people with physical access to
+//! the hardware. This is true even in cases where the program is controling
```
Comment (sstangl) [inline library/std/src/os/unix/io/mod.rs:60]:
```suggestion
//! the hardware. This is true even in cases where the program is controlling
```

File: library/std/src/os/unix/io/mod.rs [hunk 2]
```diff
@@ -44,6 +44,26 @@
 //! Like boxes, `OwnedFd` values conceptually own the resource they point to,
 //! and free (close) it when they are dropped.
 //!
+//! ## What about `/proc/self/mem` and similar OS features?
+//!
+//! Some platforms have a feature known as `/proc/self/mem`, which is a
```
Comment (joshtriplett) [inline library/std/src/os/unix/io/mod.rs:49]:
```suggestion
//! Some platforms have special files, such as `/proc/self/mem`, which is a
```

File: library/std/src/os/unix/io/mod.rs [hunk 3]
```diff
@@ -44,6 +44,26 @@
...
+//! filesystem path that can be opened, producing a file descriptor that, when
+//! read from or written to, reads and writes the process's memory. These reads
+//! and writes happen outside the control of the Rust compiler, so they do not
+//! uphold Rust's memory safety guarantees.
+//!
+//! Does this mean that all APIs that might allow `/proc/self/mem` to be opened
+//! and read from or written to must be `unsafe`? No. Rust's safety guarantees
+//! only cover what the program itself can do, and not what entities outside
+//! the program can do to it. `/proc/self/mem` is considered to be such an
+//! external entity, along with debuggers and people with physical access to
```
Comment (joshtriplett) [inline library/std/src/os/unix/io/mod.rs:59]:
```suggestion
//! external entity, along with debugging interfaces, and people with physical access to
```

**Comment by joshtriplett** [comment]
@rfcbot merge

**Review by jhpratt — APPROVED** [review]

File: library/std/src/os/unix/io/mod.rs [hunk 4]
```diff
@@ -44,6 +44,26 @@
...
+//! ## What about `/proc/self/mem` and similar OS features?
+//!
+//! Some platforms have special files, such as `/proc/self/mem`, which is a
+//! filesystem path that can be opened, producing a file descriptor that, when
+//! read from or written to, reads and writes the process's memory. These reads
+//! and writes happen outside the control of the Rust compiler, so they do not
+//! uphold Rust's memory safety guarantees.
+//!
+//! Does this mean that all APIs that might allow `/proc/self/mem` to be opened
+//! and read from or written to must be `unsafe`? No. Rust's safety guarantees
```
Comment (m-ou-se) [inline library/std/src/os/unix/io/mod.rs:56]: I think it's easier to read as statements, instead of questions. The user who is reading this has questions, but the documentation should provide answers, not ask more questions. ^^

```suggestion
//! ## `/proc/self/mem` and similar OS features
//!
//! Some platforms have special files, such as `/proc/self/mem`, which is a
//! filesystem path that can be opened, producing a file descriptor that, when
//! read from or written to, reads and writes the process's memory. These reads
//! and writes happen outside the control of the Rust compiler, so they do not
//! uphold Rust's memory safety guarantees.
//!
//! This does not mean that all APIs that might allow `/proc/self/mem` to be opened
//! and read from or written to must be `unsafe`. Rust's safety guarantees
```

**Comment by JakobDegen** [comment]
A couple thoughts: First, questions about the rules for interactions between these OS interfaces and normal Rust code seem very much like a T-Lang/UCG issue, not a T-Libs one. As the proposal here correctly points out, this is not a concern that is unique to the `File` API (or indeed any API).

Second, I think there are two separate issues here, and we should probably take care to separate them:

 1. It is clearly possible to cause UB via `File` access to `/proc/self/mem`. Why then does that API not have to be `unsafe`? The proposed answer is that it is because that is somehow "external" but this seems unsatisfying at the very least. It would imply that `dlclose` also does not have to be `unsafe` - after all, that is also "just" a tool for telling the OS to mangle your process's memory.

 2. If code does go and use these APIs, what are the rules governing the behavior of the resulting program? If a debugger/`/proc/mem/self`/whatever goes and writes to some memory in the process, is that UB? Will the result of that write be consistently reflected in subsequent code?

Imo each of these is sufficiently subtle to deserve its own full discussion

**Comment by sunfishcode** [comment]
> It is clearly possible to cause UB via File access to /proc/self/mem. Why then does that API not have to be unsafe? The proposed answer is that it is because that is somehow "external" but this seems unsatisfying at the very least. It would imply that dlclose also does not have to be unsafe - after all, that is also "just" a tool for telling the OS to mangle your process's memory.

`dlclose` doesn't depend on how end users have configured their environment. It is documented to do things tantamount to calling `free` on raw pointers, and that makes it `unsafe`.

End users can mount all kinds of things in their filesystem namespaces, including FUSE mount points where system calls invoke arbitrary side effects. Rust can't make guarantees about what `File::open` on an arbitrary path might do.

> If code does go and use these APIs, what are the rules governing the behavior of the resulting program? If a debugger//proc/mem/self/whatever goes and writes to some memory in the process, is that UB? Will the result of that write be consistently reflected in subsequent code?

It's UB.

> Imo each of these is sufficiently subtle to deserve its own full discussion

There have been discussions over the years, in https://github.com/rust-lang/rust/issues/32670, https://github.com/rust-lang/rfcs/pull/3128, and https://users.rust-lang.org/t/proposal-to-organize-bounty-for-linux-kernel-patch-to-restrict-access-to-processs-own-memory-via-file-system/58303.

**Comment by JakobDegen** [comment]
> `dlclose` doesn't depend on how end users have configured their environment. It is documented to do things tantamount to calling free on raw pointers, and that makes it unsafe.

So is writing to `/proc/self/mem`.

> End users can mount all kinds of things in their filesystem namespaces, including FUSE mount points where system calls invoke arbitrary side effects. Rust can't make guarantees about what File::open on an arbitrary path might do.

If the issue is that we can't make this guarantee, we are also admitting this is a problem, which I do not think is correct. We should instead be trying to find some reasonable set of rules that describes when people writing such APIs need to mark their functions as `unsafe` - and it should be a consequence of these rules that `dlclose` needs to be `unsafe` but `File::open` does not.

> It's UB.

It's most definitely not unconditional UB, because eg `asm!` exists and can be used soundly. It *might* be UB, but there are some thorny questions about this - for example, the AM "knows" what calls to the allocator are and will in some cases remove them, so an allocator call writing to some pointer found in a global is UB. On the other hand, `asm!` blocks are always opaque to the AM, and so they are free to write to a pointer stored in a global. It's not clear to me which side of this line a `write` call falls on - is the compiler allowed to assume that it does what it normally does? (Again, I think this PR is probably not the right place to have this discussion, I'm just trying to demonstrate the complexity)

**Comment by sunfishcode** [comment]
> > End users can mount all kinds of things in their filesystem namespaces, including FUSE mount points where system calls invoke arbitrary side effects. Rust can't make guarantees about what File::open on an arbitrary path might do.
> 
> If the issue is that we can't make this guarantee, we are also admitting this is a problem, which I do not think is correct. We should instead be trying to find some reasonable set of rules that describes when people writing such APIs need to mark their functions as `unsafe` - and it should be a consequence of these rules that `dlclose` needs to be `unsafe` but `File::open` does not.

The rules are roughly:
 - dereferencing raw pointers [etc.](https://doc.rust-lang.org/nomicon/what-unsafe-does.html) is unsafe
 - accessing the process' filesystem namespace is safe
 - external entities accessing Rust's memory in ways that effectively violate Rust's aliasing rules induce UB.
 - Rust does not attempt to prevent UB induced by external entities.
 - The Linux `procfs` driver is an external entity.

Since the initial report 6 years ago, the alternatives that have been proposed for fixing this problem in `std::fs` are:
 - Make all functions that take `Path`s unsafe.
 - Make all functions that take `Path`'s fail if they can't prove that they aren't accessing `/proc/self/mem` or any other file that Rust believes might allow the program to tell an external entity to induce UB. We could check path strings up front, open the path and do `fstat` and additional system calls to inspect it; either way, they add a level of overhead which would be difficult to justify imposing on all `std::fs` users, fix only a subset of the problem, break legitimate use cases, or combinations thereof.

At this time, I'm not aware of any other ideas for fixing this within `std::fs`. If you have any, please suggest them.

> It's most definitely not unconditional UB, because eg `asm!` exists and can be used soundly. It _might_ be UB, but there are some thorny questions about this - for example, the AM "knows" what calls to the allocator are and will in some cases remove them, so an allocator call writing to some pointer found in a global is UB (or maybe not?). On the other hand, `asm!` blocks are always opaque to the AM, and so they are free to write to a pointer stored in a global. It's not clear to me which side of this line a `write` call falls on - is the compiler allowed to assume that it does what it normally does? (Again, I think this PR is probably not the right place to have this discussion, I'm just trying to demonstrate the complexity)

What does "the AM" refer to?

I don't understand what you're saying here. `asm!` isn't permitted to violate Rust's aliasing rules; it can't dereference any pointer that compiled Rust code can't.

**Comment by JakobDegen** [comment]
> The rules are roughly: [...]

If these rules, as written, are the ones we want, then its perfectly fine for `dlclose` to be safe. (unless you want to argue that `ld` is not an external entity, but that seems pretty far-fetched). I don't think these are the rules we should have, and even if it is, it seems entirely unclear from what has been said so far that this is the decision which this FCP is making.

> What does "the AM" refer to?

Abstract machine

> I don't understand what you're saying here. asm! isn't permitted to violate Rust's aliasing rules; it can't dereference any pointer that compiled Rust code can't.

Indeed, it is not, which is why `asm!` *can* be unsound. But one can also write `asm` which does not break the aliasing rules, and such `asm` is sound. My point was that the same is possibly true of writes to `/proc/self/mem`. Of course they *might* violate the aliasing rules, but what if they don't - are they still UB? Probably not, but this hasn't really been discussed.

Again, I know I've said this plenty of times now, but I'll say it again: I am not trying to shill my position on either of these matters here. My point is that they are potentially non-obvious, and that I am unclear on what this FCP is actually deciding and how it relates to what I brought up.

**Comment by sunfishcode** [comment]
> If these rules, as written, are the ones we want, then its perfectly fine for `dlclose` to be safe. (unless you want to argue that `ld` is not an external entity, but that seems pretty far-fetched). I don't think these are the rules we should have, and even if it is, it seems entirely unclear from what has been said so far that this is the decision which this FCP is making.

Well, `ld` is not an external entity :-). I realize I haven't formally defined the term, but informally, `ld` is a very different kind of thing. Object files that will contribute code to the program making undefined references to symbols and the linker satisfying them with definitions at build time is a different category from the program doing I/O and communicating with entities that exist at runtime.

> Indeed, it is not, which is why `asm!` _can_ be unsound. But one can also write `asm` which does not break the aliasing rules, and such `asm` is sound. My point was that the same is possibly true of writes to `/proc/self/mem`. Of course they _might_ violate the aliasing rules, but what if they don't - are they still UB? Probably not, but this hasn't really been discussed.

If I understand correctly, this turns out to be equivalent asking "what if there's a write to /proc/self/mem, but it doesn't do anything that couldn't be done with `copy_nonoverlapping`?" That is an interesting question. I could imagine making an argument that that isn't UB. However, I don't think it's important to answer that question here&mdash;what's important here is just that writing to /proc/self/mem *can* have UB, and that Rust doesn't attempt to prevent it.

> Again, I know I've said this plenty of times now, but I'll say it again: I am not trying to shill my position on either of these matters here. My point is that they are potentially non-obvious, and that I am unclear on what this FCP is actually deciding and how it relates to what I brought up.

I don't yet have a clear sense of your concerns. Would you be able to write down what you think the rules should be, even if in an informal way?

**Comment by JakobDegen** [comment]
> the linker satisfying them with definitions at build time

But that's my point, dynamic linking is not done at built time, it's done at run time.

> I don't yet have a clear sense of your concerns. Would you be able to write down what you think the rules should be, even if in an informal way?

It's not a matter of rules, it's more a matter of me being unclear what is actually being FCPd, and especially why this is a libs issue. If the only thing being FCPd were "`File` will not try and protect you from accessing eg `/proc/self/mem`" that would be a different story, but I don't think that's what's going on here. Specifically this language:

 > Rust's safety guarantees only cover what the program itself can do, and not what entities outside the program can do to it. `/proc/self/mem` is considered to be such an external entity, along with debugging interfaces, and people with physical access to the hardware.

This seems to make a pretty strong claim about what it means for an API to be sound, and I want to make sure that this claim goes through the proper procedures instead of being unintentionally stabilized as part of a PR that looks like it's just about `/proc/self/mem`

**Comment by ChrisDenton** [comment]
It does seem like we're starting with the premise:

* `File` creation is safe.
* `File` operations are safe (e.g. read/write)
* Therefore operations on /proc/self/mem are not `unsafe` even though they may not be sound

And then we're working backward to try to find a way to justify this stance. I'm not entirely convinced we can. Other than to say that this is a special case because we really do not want to have the user wrap all file I/O in unsafe blocks.

In short, I don't think we should attempt to apply any justification more broadly. It's tenuous as it is.

**Comment by sunfishcode** [comment]
@JakobDegen

> > the linker satisfying them with definitions at build time
>
> But that's my point, dynamic linking is not done at built time, it's done at run time.

With dynamic linking, the dynamic linker isn't responsible for the safety of the libraries it's asked to load.

> > Rust's safety guarantees only cover what the program itself can do, and not what entities outside the program can do to it. `/proc/self/mem` is considered to be such an external entity, along with debugging interfaces, and people with physical access to the hardware.
>
> This seems to make a pretty strong claim about what it means for an API to be sound, and I want to make sure that this claim goes through the proper procedures instead of being unintentionally stabilized as part of a PR that looks like it's just about `/proc/self/mem`

I don't think the first sentence is a strong claim, because every computer programming language ever created effectively depends on this kind of claim. "The behavior of the program after someone has attached gdb to the process and performed arbitrary memory operations" is not something that any language has ever defined.

The second sentence does get specific in a non-trivial way. There's a boundary, and we don't have an explicit definition for it. But I do think there's an intuitive boundary, which I'll attempt to illustrate in more detail here:

A human being with a soldering iron and physical access to the hardware is external.

Bitwise XOR is internal.

gdb is external.

`copy_nonoverlapping` is internal.

`read` is half-and-half. The source of the data is external, and reading from it may cause it to perform arbitrary outside-world side effects. But there is also an internal part: `read` itself will always just write data in-bounds into the provided buffer.

Those outside-world side effects could be anything, including attaching to the calling process with gdb and writing beyond the bounds of the provided buffer. But, we don't mention that in `read`'s documentation. We don't say that it's the `read` that does that, in any language. We don't make `read` unsafe in any memory-safe programming language. We don't say that `read` has an array bounds violation bug. `read` just requests bytes, and writes them in-bounds in the buffer. This interpretation is common to all computer programming languages that have a `read` function that reads bytes from a stream into a buffer.

`open` is similarly half-and-half, though it is a little more interesting. There are a few different places where one might reasonably place the boundary between internal and external, however Rust's `std::fs` and most popular languages' standard libraries today are organized around an ambient filesystem namespace which effectively determines their stance here. Regardless, the point is that there is a boundary somewhere. A FUSE filesystem daemon has to be outside that boundary, wherever the boundary is, because it's an arbitrary program capable of arbitrary side effects. What this means is that there can be things mounted in the filesystem namespace that, when opened, provide access to I/O devices like FUSE filesystems that invoke arbitrary outside-world side effects. As with `read`, we don't say that this makes `open` unsafe. We don't document the possibility of undefined behavior in `open` in any programming language that has an `open`, even though it *might* open a FUSE filesystem and the FUSE daemon might perform arbitrary outside-world side effects. So there is a boundary. And there are things outside the boundary. And to be sure, this is all before we start thinking about `procfs`.

So now let's look at `procfs`. My second sentence above is claiming that when you read from or write to `/proc/self/mem`, the way it works is that there's an internal side to `read` and `write` which just do their job and have fully defined behavior, as we described earlier, and separately, there's an external side where the Linux `procfs` driver has the outside-world side effect of using kernel magic to reach into the process and manipulate the memory. This isn't `read`'s or `write`'s or even `open`'s fault. It doesn't oblige them to be unsafe. It doesn't oblige them to document this possibility. It's just one of the infinitely-many things outside-world things that can happen when a program does I/O.

This is true even if a program is written to deliberately open /proc/self/mem, knowing full well what it does, to deliberately corrupt its own memory. Rust doesn't take responsibility for that, just as it doesn't take responsibility for programs deliberately spawning gdb and sending it commands to attach to their own process and corrupt their own memory. This is how I propose we explain how Rust has worked since the beginning. I'll additionally claim here that this is effectively how all general-purpose programming languages ported to Linux have ever worked.

@ChrisDenton

> In short, I don't think we should attempt to apply any justification more broadly. It's tenuous as it is.

With my explanation above, I have attempted to show that we can do it, and that it's not tenuous.

**Comment by sunfishcode** [comment]
It turns out the implementation of `/proc/self/mem` [calls `access_remote_vm`](https://github.com/torvalds/linux/blob/7609385337a4feb6236e42dcd0df2185683ce839/fs/proc/base.c#L867). Unlike `copy_from_user` or `copy_to_user` which are the normal way for the kernel to access memory in userspace, nothing else in the entire kernel calls `access_remote_vm` except for one other thing: `ptrace`&mdash;the API to implement debuggers such as gdb. So this analogy we have here about how `/proc/self/mem` is similar to attaching a debugger is reflected in the source code for `/proc/self/mem` itself.

**Comment by JakobDegen** [comment]
@sunfishcode that seems like a reasonable set of rules. It's a little unclear to me how `dlclose` fits into that, but if you can argue that it does, I'd be happy.

But is that set of rules what this FCP is approving? Seems unlikely, since it wasn't even written down before now. Maybe it was just unclear to me though.

**Comment by ChrisDenton** [comment]
> Making it unsafe would be coherent, but on a practical level, it would likely do more harm than good.

This I very much agree with. Rust's stance can be fully justified on pragmatic grounds.

What I'm less clear is on is the proposed "internal" and "external" dichotomy. If I understand you right, a hypothetical `libc::corrupt_my_memory` function would be internal but a `File::open("/special/corrupt_my_memory")` would be external?

Also, if Rust could encapsulate "person with a soldering iron" then surely it should do so? The problem with that is a practical one, not a philosophical objection to doing so.

**Comment by sunfishcode** [comment]
> What I'm less clear is on is the proposed "internal" and "external" dichotomy. If I understand you right, a hypothetical `libc::corrupt_my_memory` function would be internal but a `File::open("/special/corrupt_my_memory")` would be external? Even if they used the same underlying mechanism?

The underlying mechanism matters. If it's doing `File::open`, users have the option to control what it can do by running it in a sandbox.

> Also, if Rust could encapsulate "person with a soldering iron" then surely it should do so? The problem with that is a practical one, not a philosophical objection to doing so.

"Is there a boundary?" is philosophical while "where is the boundary?" is practical.

**Comment by cuviper** [comment]
> a hypothetical `libc::corrupt_my_memory` function

There is [`libc::process_vm_writev`](https://docs.rs/libc/latest/libc/fn.process_vm_writev.html), but at least that's `unsafe`, like any FFI. However, `nix` has a [safe wrapper](https://docs.rs/nix/latest/nix/sys/uio/fn.process_vm_writev.html), which is probably a mistake.

**Comment by sunfishcode** [comment]
In theory, `process_vm_writev` could go either way. On one hand, users could stop it with `seccomp` or by withholding `PTRACE_MODE_ATTACH_REALCREDS`, so one could reasonably classify the access as "external" and call it safe. On the other hand, it's not widely used, and it's quite easy to get it to invoke UB in the calling process, so it seems practical and more desirable to classify it as unsafe.

**Comment by JakobDegen** [comment]
I unfortunately didn't get a chance to respond before, and not sure if it's worth doing it now, but I guess I will anyway.

> Rust isn't alone here; lots of programming languages have effectively the exact same rules, yet as far as I know, none of them document the rules as such. I think there's a widespread assumption that this overall framework is just how computers are understood to work and it doesn't need to be stated.

Not at all. There's no other language I know of that attempts to make a formal distinction between "sound" and "unsound," and so no other language has to deal with this issue. I think I've actually made a mistake in not emphasizing this point enough in this conversation. "Sound" and "unsound" are not (supposed to be) vague ideas that represent some philosophical goals, they are concrete concepts that we hope to tie to a formal definition. I don't mean to be rude here, but it frankly seems like this is being hand-waved away both in this conversation and in the PR itself, and I think that is a distinctly dangerous road to go down.

I went back and re-read your proposed rules, and honestly I'm not quite sure why I was happy with them the first time around - I don't agree with the distinctions, and I especially don't believe that arguing these kinds of points from a list of examples is a way forward. I have some alternative ideas in mind, and would be happy to discuss more on Zulip instead of here.

> So perhaps what we're seeing here and in other places where people keep asking questions about this, is that not everyone makes the same assumptions or thinks about this the same way. In that case, perhaps the path forward looks like:
>
> - Assuming this FCP completes here, we merge this PR, establishing that this is T-libs' official position, at least in the eyes of people who share the overall intuition.
> - Then, someone could write up the rules above in a form that could be added to the language-level documentation, perhaps in the Nomicon, because I think this relates to your questions about why this isn't a T-lang/UCG question.
> - Then, we could update the library documentation here to reference that language documentation.
>
> How does that sound?

Honestly, about as backwards as could be. "Stabilize that this API is sound first and figure out some rules under which that makes sense later" is not at all how these things should work. When we add the language level documentation, what choice will be left for T-Lang and T-types? What if they (like me) don't agree that we can just call this sound? It's not like they'll be able to say no anymore.

I suppose it's on me that I failed to ask for this to be nominated for T-Types earlier

**Comment by ChrisDenton** [comment]
I mean, it's not set in stone until it reaches stable? I think you can still discuss this on Zulip and ask t-libs if you have reasons they might want to reconsider.

I personally don't want to take issue with `/proc/self/mem` being safe (even if it's far from ideal). I am however uncomfortable with the broad reasoning given in this PR, particularly how it might be applied to other cases that can be encapsulated without the same practical problems of making `File` unsafe to use.

**Comment by JakobDegen** [comment]
Yeah, agreed. If this was just documenting that std doesn't care about protecting you from `/proc/self/mem` we'd be fine.

That's a good point about us having until this reaches stable. I might bring this up with T-types on zulip then (as they are the proper team to be making this decision). Later though, it is bed time for me right now

**Comment by SimonSapin** [comment]
For anyone else looking, this doc-comment ends up visible at: https://doc.rust-lang.org/nightly/std/os/unix/io/index.html#procselfmem-and-similar-os-features
